building/qubes-r3-building.md

```diff
@@ -24,36 +38,80 @@ $ wget https://keys.qubes-os.org/keys/qubes-master-signing-key.asc
 $ gpg --import qubes-master-signing-key.asc 
 $ gpg --edit-key 36879494
 # Verify fingerprint!, set trust to *ultimate*
+# (Typical sequence is fpr, trust, 5, q)
 $ wget https://keys.qubes-os.org/keys/qubes-developers-keys.asc
 $ gpg --import qubes-developers-keys.asc
 ~~~
 
-Note we do *not* relay above on the security of our server (keys.qubes-os.org) nor the connection (ssl, cert) -- we only rely on you getting the Qubes Master Signing Key fingerprint *somehow* and ensure they match!
+And if you will be building the Whonix templates:
+
+~~~
+$ wget https://github.com/QubesOS/qubes-builder-debian/blob/master/keys/whonix-developer-patrick.asc
+$ gpg --import whonix-developer-patrick.asc
+~~~
+
+**Note** In the above process, we do *not* rely on the security of our server (keys.qubes-os.org) nor the connection (ssl, cert) -- we only rely on you getting the Qubes Master Signing Key fingerprint *somehow* and ensuring they match!
+Likewise, the Whonix signing key is available from multiple sources.
 
-Now lets bootstrap the builder. Unfortunately the builder cannot verify itself (the classic Chicken and Egg problem), so we need to verify the signature manually:
+Now let's bootstrap the builder. Unfortunately, the builder cannot verify itself (the classic Chicken and Egg problem), so we need to verify the signature manually:
 
 ~~~
 $ git clone git://github.com/QubesOS/qubes-builder.git
+$ mkdir qubes-builder/keyrings
+$ mkdir qubes-builder/keyrings/git
+$ cp .gnupg/pubring.gpg qubes-builder/keyrings/git/
+$ cp .gnupg/trustdb.gpg qubes-builder/keyrings/git/
 $ cd qubes-builder
-$ git describe --exact-match HEAD
-<some tag>
-$ git tag -v <some tag>
+$ git tag -v `git describe`
+~~~
+
+Assuming the verification went fine, we're good to go with all the rest without ever thinking more about verifying digital signatures on all the rest of the components.
+The builder will do that for us for each component, every time we build, even for all auxiliary files (e.g. Xen or Linux kernel sources).
+
+Let's configure the builder first (see Note at bottom if you would prefer to manually configure):
+
+~~~
+$ ./setup
+# Select 3.2 or 4.0 for version
+# Stable
+# No (we want a full build)
+# Select builder-fedora, builder-debian, template-whonix, mgmt-salt
+# Yes (to download)
+# Select fc26, stretch, whonix-gateway, whonix-workstation (for the currently shipping templates)
 ~~~
 
-Assuming the verification went fine, we're good to go with all the rest without ever thinking more about verifying digital signatures on all the rest of the components, as the builder will do that for us, for each component, every time we, even for all aux files (e.g. Xen or Linux kernel sources).
+Once the download is complete, continue the build process with:
+
+~~~
+$ make install-deps
+$ make get-sources
+~~~
```

Comment:
setup script will take care about importing Whonix signing key (there will be a prompt for that).
But if going with manual builder configuration, this is a good place for importing that key - it is already downloaded to `qubes-src/builder-debian/keys/whonix-developer-patrick.asc`, like this:
```
gpg --homedir keyrings/git --import qubes-src/builder-debian/keys/whonix-developer-patrick.asc
```
You may need to execute `make get-sources` again to resume the operation (if any component failed because of missing key).

Reply:
That's why I'm trying to pre-stage it in the previous section, before we start copying keyrings around and before the user starts hitting errors related to it being missing.

Reply:
Hmm, I wonder if this step is required at all, since I review updates to qubes-template-whonix and add my tag there too. It is probably only needed for development builds directly from Patrick's repository.